### PR TITLE
🔧 Allow `get_results` to find `Experiment.results` by app name, and optionally version

### PR DESCRIPTION
**Pull request by juanesarango**
## Goal

There's a limitation on the current implementation of `AbstractApplication.get_results`, as it matches only by application's primary key. There's a lot of use cases, where you want to find results from same application even if from an older version.

## Context of the problem

We use `get_results` often to fetch results from `get_dependencies`, which are limited to strict application primary keys.
Let’s say you have these applications, where `ANNOT` relies on `MUTECT (v1.0)` and `STRELKA (v1.0)`.

```
pk  name  version
-------------------
1   MUTECT   v1.0
2   STRELKA  v1.0
3   ANNOT    v1.0
```
If you upgrade `STRELKA` to `v2.0` you get:
```
pk  name  version
-------------------
1   MUTECT   v1.0
2   STRELKA  v1.0
3   ANNOT    v1.0
4   STRELKA  v2.0
```
In current implementation, `ANNOT` now will work only with `MUTECT (v1.0)` and `STRELKA (v2.0)`, as it needs specific Strelka primary_key  to find the results. In some scenarios, you would like to use `STRELKA (v1.0)` if results are available.

That's why in some scenarios, it would be useful to fetch results based only with the app name.

## Solution

I introduce a new attribute property to the `AbstractApplication` class, that allows to define the results needed, and these would be automatically retrieved, 🎉 **without the need to define the `get_dependencies` method** 🎉. In fact, it will override it.

### With Fixed Version
```python
@cached_property
def dependencies_results(self):
    return [
        {
            "app": apps.StrelkaV2(), 
            "result": "snvs", 
            "name": "strelka_snvs"
        },
        {
            "app": apps.MutectV1(), 
            "result": "snvs", 
            "name": "strelka_snvs"
        },
    ]
```

or it's equivalent, specifying `app_name` and `app_version`:

```python
@cached_property
def dependencies_results(self):
    return [
        {
            "app_name": apps.StrelkaV2().NAME, 
            "app_version": apps.StrelkaV2().VERSION,
            "result": "snvs", 
            "name": "strelka_snvs"
        },
        {
            "app_name": apps.MutectV1().NAME,
            "app_version": apps.MutectV1().VERSION,
            "result": "snvs", 
            "name": "strelka_snvs"
        },
    ]
```

### Without Fixed Version
As the version is optional. In this case, any result from `STRELKA (1.0)` or `STRELKA (2.0)` if available would work

```python
dependencies_results = [
    {
        "app_name": "STRELKA", 
        "result": "snvs", 
        "name": "strelka_snvs"
    },
    {
        "app_name": "MUTECT",
        "result": "snvs", 
        "name": "strelka_snvs"
    },
]
```
➕  In this last case, if  you define `version` as `any` and there's more than 1 available result, it will take the latest run. 

## Extras 

- `{ ...link: false }` the analysis of the result won't be linked to the analysis. As this creates new uniqueness conditions.

Some times, adding new dependencies to an analysis, will change the analysis **uniqueness restrictions**, causing that isabl will create a new one if same targets and references are run for the same app, as it has now different analyses dependencies. 

- `options.get_dependency_analyses_options` it's a `--fi <filter-analyses` option that uses the new `dependency_results` list to retrieve the existing analyses dependencies to define the experiments tuples.
i.e 
```python
@property
def cli_options(self):
    return options.get_dependency_analyses_option(
        self.dependencies_results
    )
```

### CODE REVIEW

**Comment by juanesarango** [comment]
These changes, have been validates in papaemmanuil isabl apps:

Running changes in `isabl_apps`:
```bash
# install isabl_apps
git clone https://github.com/papaemmelab/isabl_apps
cd isabl_apps
pip install -r requirements.txt 

# Run tests using specific isabl_cli branch
ISABL_CLI_BRANCH=improve-app-dependencies
sh tests/test-with-compose.sh
```
**Output:**
```text
=================================== short test summary info ===================================
SKIP [1] tests/test_cellranger.py:13: Transcriptome reference folder not available
SKIP [1] tests/test_data.py:55: samtools required for this test, install samtools?
SKIP [1] tests/test_data.py:65: samtools required for this test, install samtools?

Results (207.64s):
      51 passed
       3 skipped
```

**Comment by juanesarango** [comment]
The Codepatch coverage reports are failing from previous untested commits that havem't updated the coverage status of the repo.

File: .travis.yml [hunk 1]
```diff
@@ -6,6 +6,7 @@ services:
 - docker
 before_install:
 - echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
+- echo -e "machine github.com\n  login juanesarango \npassword $GH_TOKEN" > ~/.netrc
```
Comment (ddomenico) [inline .travis.yml:9]: should we change the login here from your user to an env variable? `$GH_USERNAME` perhaps?

**Review by ddomenico — APPROVED** [review]
Great update and excellent addition of tests! 💪 This is going to be super useful.

File: .travis.yml (hunk 1 above)
Comment (juanesarango) [reply .travis.yml]: Good idea

**Pull request merged**